storage/innobase/fil/fil0crypt.cc

```diff
@@ -635,36 +707,112 @@ static fil_space_crypt_t *fil_space_set_crypt_data(
 @return position on log buffer */
-
-byte *fil_parse_write_crypt_data(byte *ptr, const byte *end_ptr,
-                                 const buf_block_t *block, ulint len) {
+byte *fil_parse_write_crypt_data_v2(space_id_t space_id, byte *ptr,
+                                    const byte *end_ptr, ulint len) {
   ptr += 4;  // skip offset and len
-  const uint iv_len = mach_read_from_2(ptr + ENCRYPTION_MAGIC_SIZE);
-  ut_a(iv_len == CRYPT_SCHEME_1_IV_LEN);  // only supported
 
   if (len != KERYING_ENCRYPTION_INFO_MAX_SIZE) {
     recv_sys->set_corrupt_log();
-    return NULL;
+    return nullptr;
   }
 
   if (ptr + KERYING_ENCRYPTION_INFO_MAX_SIZE > end_ptr) {
-    return NULL;
+    return nullptr;
   }
 
   // We should only enter this function if ENCRYPTION_KEY_MAGIC_PS_V1 is set
-  ut_ad((memcmp(ptr, ENCRYPTION_KEY_MAGIC_PS_V1, ENCRYPTION_MAGIC_SIZE) == 0));
+  ut_ad((memcmp(ptr, ENCRYPTION_KEY_MAGIC_PS_V2, ENCRYPTION_MAGIC_SIZE) == 0));
   ptr += ENCRYPTION_MAGIC_SIZE;
 
-  ptr += 2;  // length of iv has been already read
+  uint type = mach_read_from_1(ptr);
+  ptr += 1;
+
+  ut_a(type == CRYPT_SCHEME_UNENCRYPTED ||
+       type == CRYPT_SCHEME_1);  // only supported
+
+  uint min_key_version = mach_read_from_4(ptr);
+  ptr += 4;
 
-  space_id_t space_id = mach_read_from_4(ptr);
+  uint key_id = mach_read_from_4(ptr);
   ptr += 4;
-  uint offset = mach_read_from_2(ptr);
-  ptr += 2;
+
+  fil_encryption_t encryption = (fil_encryption_t)mach_read_from_1(ptr);
+  ptr += 1;
+
+  fil_space_crypt_t *crypt_data = fil_space_create_crypt_data(
+      encryption, key_id, Crypt_key_operation::FETCH_OR_GENERATE_KEY);
+  /* Need to overwrite these as above will initialize fields. */
+  DBUG_ASSERT(min_key_version != ENCRYPTION_KEY_VERSION_INVALID);
+  crypt_data->min_key_version = min_key_version;
+  crypt_data->encryption = encryption;
+  memcpy(crypt_data->iv, ptr, CRYPT_SCHEME_1_IV_LEN);
+  ptr += CRYPT_SCHEME_1_IV_LEN;
+  crypt_data->encryption_rotation =
+      static_cast<Encryption_rotation>(mach_read_from_1(ptr));
+  ptr += 1;
+  uchar tablespace_key[ENCRYPTION_KEY_LEN];
+  memcpy(tablespace_key, ptr, ENCRYPTION_KEY_LEN);
+  ptr += ENCRYPTION_KEY_LEN;
+
+  if (std::search_n(tablespace_key, tablespace_key + ENCRYPTION_KEY_LEN,
+                    ENCRYPTION_KEY_LEN,
+                    0) ==
+      tablespace_key) {  // tablespace_key is all zeroes which means there is no
+                         // tablepsace in mtr log
+    crypt_data->set_tablespace_key(nullptr);
+    ptr += ENCRYPTION_KEY_LEN;
+  } else {
+    crypt_data->set_tablespace_key(tablespace_key);
+  }
+
+  /* Check is used key found from encryption plugin */
+  if (crypt_data->should_encrypt() && !crypt_data->is_key_found()) {
+    ib::error() << "Key cannot be read for space id = " << space_id;
+    recv_sys->set_corrupt_log();
+  }
+
+  /* update fil_space memory cache with crypt_data */
+  if (fil_space_t *space = fil_space_acquire_silent(space_id)) {
```

Comment:
would be nice if you can avoid assignments inside if expression..

Reply:
done